### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by rauljareno**

### CODE REVIEW

File: lib/identity/data/data_sources/rpc_identity_data_source.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  /// Retrieve last state for a given identity.
+  ///
+  /// @param [id] identity - The base58 identifier string
+  /// @param [stateContract] the ABI contract
+  ///
+  /// @returns [String] last state committed
+  Future<String> getState(String id, DeployedContract stateContract) async {
+    EnvEntity env = await _getEnvUseCase.execute();
+    Web3Client web3Client = getItSdk.get(param1: env);
```
Comment (Plumillon) [inline lib/identity/data/data_sources/rpc_identity_data_source.dart:25]: This should be injected, it's the source.

File: lib/identity/data/data_sources/rpc_identity_data_source.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import 'package:polygonid_flutter_sdk/common/domain/entities/env_entity.dart';
+import 'package:polygonid_flutter_sdk/common/domain/use_cases/get_env_use_case.dart';
+import 'package:polygonid_flutter_sdk/common/utils/uint8_list_utils.dart';
+import 'package:polygonid_flutter_sdk/assets/state.g.dart';
+import 'package:polygonid_flutter_sdk/sdk/di/injector.dart';
+import 'package:web3dart/crypto.dart';
+import 'package:web3dart/web3dart.dart';
+
+class RPCIdentityDataSource {
+  final GetEnvUseCase _getEnvUseCase;
```
Comment (Plumillon) [inline lib/identity/data/data_sources/rpc_identity_data_source.dart:13]: Need to remove this one and pass the data as param.

File: lib/identity/data/repositories/smt_repository_impl.dart [hunk 3]
```diff
@@ -204,6 +197,6 @@ class SMTRepositoryImpl implements SMTRepository {
 
   @override
   Future<Map<String, dynamic>> convertState({required TreeStateEntity state}) {
-    return Future.value(_treeStateMapper.mapTo(state));
+    return Future.value(_treeStateMapper.mapTo(state).toJson());
```
Comment (Plumillon) [inline lib/identity/data/repositories/smt_repository_impl.dart:200]: Add the `.toJson()` to the mapper.

File: lib/proof/data/data_sources/lib_pidcore_proof_data_source.dart [hunk 4]
```diff
@@ -129,7 +129,7 @@ class LibPolygonIdCoreProofDataSource {
     Map<String, dynamic>? gistProof,
     String? challenge,
     String? signature,
-    required ClaimInfoDTO credential,
+    required Map<String, dynamic> credential,
     required ProofScopeRequest request,
```
Comment (Plumillon) [inline lib/proof/data/data_sources/lib_pidcore_proof_data_source.dart:133]: The request should already be a Map<> at this point.

File: lib/proof/data/data_sources/rpc_proof_data_source.dart [hunk 5]
```diff
@@ -9,45 +9,10 @@ import 'package:polygonid_flutter_sdk/sdk/di/injector.dart';
 import 'package:web3dart/crypto.dart';
 import 'package:web3dart/web3dart.dart';
 
-class RPCDataSource {
+class RPCProofDataSource {
```
Comment (Plumillon) [inline lib/proof/data/data_sources/rpc_proof_data_source.dart:12]: `RPCIdentityDataSource` and `RPCProofDataSource` look like duplicates?

File: lib/proof/domain/entities/gist_proof_entity.dart [hunk 6]
```diff
@@ -9,6 +9,13 @@ class GistProofEntity {
   @override
   String toString() => "[GistProofEntity] {root: $root, proof: $proof}";
 
+  @override
+  Map<String, dynamic> toJson() => {
+        'root': root,
+        'proof': proof.toJson(),
+      }..removeWhere(
+          (dynamic key, dynamic value) => key == null || value == null);
```
Comment (Plumillon) [inline lib/proof/domain/entities/gist_proof_entity.dart:17]: I saw you're removing `null` at every `toJson`, don't we need the value even if it's null in some case?

File: lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart [hunk 7]
```diff
@@ -54,34 +48,38 @@ class GetAuthInputsUseCase
...
-        key: authClaimNode.hash,
-        type: TreeType.claims,
-        did: param.genesisDid,
-        privateKey: param.privateKey);
+    Map<String, dynamic> incProof = (await _smtRepository.generateProof(
+            key: authClaimNodeHash,
+            type: TreeType.claims,
+            did: param.genesisDid,
+            privateKey: param.privateKey))
+        .toJson();
```
Comment (Plumillon) [inline lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart:66]: The UC is not the one changing the format of the data.
Either you pass the entity and the mapping is done in the Data layer, or you only pass the targeted data from the entity.

File: lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart [hunk 8]
```diff
@@ -54,34 +48,38 @@ class GetAuthInputsUseCase
...
-        key: authClaimNode.hash,
-        type: TreeType.revocation,
-        did: param.genesisDid,
-        privateKey: param.privateKey);
+    Map<String, dynamic> nonRevProof = (await _smtRepository.generateProof(
+            key: authClaimNodeHash,
+            type: TreeType.revocation,
+            did: param.genesisDid,
+            privateKey: param.privateKey))
+        .toJson();
```
Comment (Plumillon) [inline lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart:73]: Same as above.

File: lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart [hunk 9]
```diff
@@ -54,34 +48,38 @@ class GetAuthInputsUseCase
...
 
     // hash of clatr, revtr, rootr
-    Map<String, dynamic> treeState = await _getLatestStateUseCase.execute(
-        param: GetLatestStateParam(
-            did: param.genesisDid, privateKey: param.privateKey));
+    Map<String, dynamic> treeState = (await _getLatestStateUseCase.execute(
+            param: GetLatestStateParam(
+                did: param.genesisDid, privateKey: param.privateKey)))
+        .toJson();
```
Comment (Plumillon) [inline lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart:79]: Same as above.

File: lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart [hunk 10]
```diff
@@ -54,34 +48,38 @@ class GetAuthInputsUseCase
...
 
-    GistProofEntity gistProof =
-        await _getGistProofUseCase.execute(param: param.genesisDid);
+    Map<String, dynamic> gistProof =
+        (await _getGistProofUseCase.execute(param: param.genesisDid)).toJson();
```
Comment (Plumillon) [inline lib/iden3comm/domain/use_cases/get_auth_inputs_use_case.dart:82]: Same as above.

**Review by Plumillon — COMMENTED** [review]
We could have a generic EntityToJsonMapper if you don't want to have multiple dedicated ones.
Even if mappers are tied to their DS and we should have one for each (if needed), we can have this general mapper in common.

File: lib/proof/data/data_sources/rpc_proof_data_source.dart (hunk 5 above)
Comment (rauljareno) [reply lib/proof/data/data_sources/rpc_proof_data_source.dart:12]: They are not, RPCIdentity calls the smart contract to get the identity state, and RPCProof calls the smart contract to get the gist (global identity state) proof, I'm still having doubts if GistProof should be in identity or in proof, for now I think its place is in proof part as it's a proof and we are using it for generating proofs also.

**Pull request closed without merging**